src/uct/base/uct_iface.c

```diff
@@ -66,10 +69,48 @@ ucs_status_t uct_iface_set_am_handler(uct_iface_h tl_iface, uint8_t id,
 
     if (cb == NULL) {
         uct_iface_set_stub_am_handler(iface, id);
+        return UCS_OK;
+    }
+
+    /* flags != 0 */
+    if (flags == 0) {
+        ucs_debug("am cb flags must not be empty");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    status = uct_iface_query(tl_iface, &attr);
```

Comment:
at least we need to check and fail the case when sync callback is set on transport that supports only async ones.
Hopefully setting am handler is not done in the critical path and performance is not an issue here.